packages/drivers/odsp-driver/src/prefetchLatestSnapshot.ts

```diff
@@ -0,0 +1,207 @@
+    IPersistedCache,
+    ISnapshotOptions,
+    OdspResourceTokenFetchOptions,
+    snapshotKey,
+    TokenFetcher,
+} from "@fluidframework/odsp-driver-definitions";
+import { ChildLogger, PerformanceEvent, TelemetryLogger } from "@fluidframework/telemetry-utils";
+import { LocalPersistentCache } from "./odspCache";
+import { ISnapshotCacheValue } from "./odspDocumentStorageManager";
+import {
+    createOdspLogger,
+    fetchAndParseAsJSONHelper,
+    getOdspResolvedUrl,
+    getWithRetryForTokenRefresh,
+    IOdspResponse,
+} from "./odspUtils";
+import { IOdspSnapshot } from "./contracts";
+import { evalBlobsAndTrees, toInstrumentedOdspTokenFetcher } from "./odspUtils2";
+
+/**
+ * Function to prefetch the snapshot and cached it in the persistant cache, so that when the container is loaded
+ * the cached latest snapshot could be used and removes the network call from the critical path.
+ * @param resolvedUrl - Resolved url to fetch the snapshot.
+ * @param getStorageToken - function that can provide the storage token for a given site. This is
+ *  is also referred to as the "VROOM" token in SPO.
+ * @param persistedCache - Cache to store the fetched snapshot.
+ * @param logger - Logger to have telemetry events.
+ * @param hostSnapshotFetchOptions - Options to fetch the snapshot if any. Otherwise default will be used.
+ * @returns - True if the snapshot is cached, false otherwise.
+ */
+export async function prefetchLatestSnapshot(
+    resolvedUrl: IResolvedUrl,
+    getStorageToken: TokenFetcher<OdspResourceTokenFetchOptions>,
+    persistedCache: IPersistedCache = new LocalPersistentCache(),
+    logger: ITelemetryBaseLogger,
+    hostSnapshotFetchOptions: ISnapshotOptions | undefined,
+): Promise<boolean> {
+    const odspLogger = createOdspLogger(ChildLogger.create(logger, "PrefetchSnapshot"));
+    const odspResolvedUrl = getOdspResolvedUrl(resolvedUrl);
+    const snapshotUrl = odspResolvedUrl.endpoints.snapshotStorageUrl;
+    let snapshotOptions: ISnapshotOptions = {
+        deltas: 1,
+        channels: 1,
+        blobs: 2,
+    };
+
+    if (hostSnapshotFetchOptions) {
+        snapshotOptions = {...snapshotOptions, ...hostSnapshotFetchOptions};
+    }
+
+    const storageTokenFetcher = toInstrumentedOdspTokenFetcher(
+        odspLogger,
+        odspResolvedUrl,
+        getStorageToken,
+        true /* throwOnNullToken */,
+    );
+
+    const snapshotCacheEntry: ICacheEntry = {
+        type: snapshotKey,
+        key: "",
+        file: {
+            resolvedUrl: odspResolvedUrl,
+            docId: odspResolvedUrl.hashedDocumentId,
+        },
+    };
+
+    return PerformanceEvent.timedExecAsync(
+        odspLogger,
+        { eventName: "PrefetchLatestSnapshot" },
+        async (event: PerformanceEvent) => {
+            let attempts = 1;
+            const success = await getWithRetryForTokenRefresh(async (tokenFetchOptions) => {
+                if (tokenFetchOptions.refresh) {
+                    attempts = 2;
```

Comment:
I'm not following the logic (maybe add a comment?). Why does attempts get set to 2 here?

Reply:
I'd do (on line 100)
attempts : tokenFetchOptions.refresh ? 2 : 1
That would be more descriptive (and shorter).

Reply:
@vladsud scope of tokenFetchOptions is not there to use it directly in the end event. So I added some description.